include/bluetooth/audio/has.h

```diff
@@ -0,0 +1,370 @@
+/*
+ * Copyright (c) 2022 Codecoup
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_BLUETOOTH_AUDIO_HAS_H_
+#define ZEPHYR_INCLUDE_BLUETOOTH_AUDIO_HAS_H_
+
+/**
+ * @brief Hearing Access Service (HAS)
+ *
+ * @defgroup bt_has Hearing Access Service (HAS)
+ *
+ * @ingroup bluetooth
+ * @{
+ *
+ * The Hearing Access Service is used to identify a hearing aid and optionally
+ * to control hearing aid presets. This API implements the server functionality.
+ *
+ * [Experimental] Users should note that the APIs can change as a part of
+ * ongoing development.
+ */
+
+#include <stdint.h>
+#include <sys/util.h>
+#include <bluetooth/conn.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#if defined(CONFIG_BT_HAS)
+#define BT_HAS_PRESET_CNT CONFIG_BT_HAS_PRESET_CNT
+#else
+#define BT_HAS_PRESET_CNT 0
+#endif /* CONFIG_BT_HAS */
+
+#define BT_HAS_PRESET_NAME_MIN 1
+#define BT_HAS_PRESET_NAME_MAX 40
+
+/** @brief Opaque Hearing Access Service object. */
+struct bt_has;
+
+/** Preset Properties values */
+enum {
+	/** No properties set */
+	BT_HAS_PROP_NONE = 0,
+
+	/** Preset name can be written by the client */
+	BT_HAS_PROP_WRITABLE = BIT(0),
+
+	/** Preset availability */
+	BT_HAS_PROP_AVAILABLE = BIT(1),
+};
+
+/** @brief Register structure for preset. */
+struct bt_has_preset_register_param {
+	/** Preset index */
+	uint8_t idx;
+	/** Preset properties */
+	uint8_t properties;
+	/** Preset name */
+#if defined(CONFIG_BT_HAS_PRESET_NAME_DYNAMIC)
+	char name[BT_HAS_PRESET_NAME_MAX];
+#else
+	const char *name;
+#endif /* CONFIG_BT_HAS_PRESET_NAME_DYNAMIC */
+};
+
+/** @brief Preset operations structure. */
+struct bt_has_preset_ops {
+	/**
+	 * @brief Preset set active callback
+	 *
+	 * Once the preset becomes active, the bt_has_preset_active_set shall
+	 * be called to notify all the clients.
```

Comment:
This is server only. The active_set callback is called then remote HAS Client attempts to change the preset. If the operation might be not possible to perform at the moment (Implementation specific) the application can return -EBUSY to simply reject the request.

But once the Preset becomes active, the application shall call `bt_has_preset_active_set`; the method used to inform all of the HAS Clients about currently used preset.

The Control Point operations is asynchronous, thus when Client sends Control Point ATT Write Request to Set Active Preset and gets ATT Write Response with error == 0 it cannot assume the active preset changed.

The specification says: "If this operation causes the value of the Active Preset Index characteristic to change, then the server shall notify clients of the new Active Preset Index value (see Section 3.3)."

So once the Server change the preset it will notify. But is something will fail - e.g. DSP re-configuration the notification will not be sent.

Reply:
@MariuszSkamra The downside of this design is of course that HAS cannot be qualified without an application on top that is responsible for sending the notifications. This is the issue we discussed in the last F2F. 

w.r.t. my question: Would it be possible for HAS to send the notification autonomously if the return code is 0?

Reply:
OK